x/auth/account.go

```diff
@@ -121,12 +139,339 @@ func (acc *BaseAccount) SetSequence(seq uint64) error {
+// sorted.
+func (bva BaseVestingAccount) spendableCoins(vestingCoins sdk.Coins) sdk.Coins {
+	var spendableCoins sdk.Coins
+	bc := bva.GetCoins()
+
+	j, k := 0, 0
+	for _, coin := range bc {
+		// zip/lineup all coins by their denomination to provide O(n) time
+		for j < len(vestingCoins) && vestingCoins[j].Denom != coin.Denom {
+			j++
+		}
+		for k < len(bva.DelegatedVesting) && bva.DelegatedVesting[k].Denom != coin.Denom {
+			k++
+		}
+
+		baseAmt := coin.Amount
+
+		vestingAmt := sdk.ZeroInt()
+		if len(vestingCoins) > 0 {
+			vestingAmt = vestingCoins[j].Amount
+		}
+
+		delVestingAmt := sdk.ZeroInt()
+		if len(bva.DelegatedVesting) > 0 {
+			delVestingAmt = bva.DelegatedVesting[k].Amount
+		}
+
+		// compute min((BC + DV) - V, BC) per the specification
+		min := sdk.MinInt(baseAmt.Add(delVestingAmt).Sub(vestingAmt), baseAmt)
+		spendableCoin := sdk.NewCoin(coin.Denom, min)
+
+		if !spendableCoin.IsZero() {
+			spendableCoins = spendableCoins.AddCoinByDenom(spendableCoin)
+		}
+	}
+
+	return spendableCoins
+}
+
+// trackDelegation tracks a delegation amount for any given vesting account type
+// given the amount of coins currently vesting. It returns the resulting base
+// coins.
+//
+// CONTRACT: The account's coins, delegation coins, vesting coins, and delegated
+// vesting coins must be sorted.
+func (bva *BaseVestingAccount) trackDelegation(vestingCoins, amount sdk.Coins) sdk.Coins {
+	bc := bva.GetCoins()
+
+	i, j, k := 0, 0, 0
+	for _, coin := range amount {
+		// zip/lineup all coins by their denomination to provide O(n) time
+		for i < len(bc) && bc[i].Denom != coin.Denom {
+			i++
+		}
+		for j < len(vestingCoins) && vestingCoins[j].Denom != coin.Denom {
+			j++
+		}
+		for k < len(bva.DelegatedVesting) && bva.DelegatedVesting[k].Denom != coin.Denom {
+			k++
+		}
+
+		baseAmt := sdk.ZeroInt()
+		if len(bc) > 0 {
+			baseAmt = bc[i].Amount
+		}
+
+		vestingAmt := sdk.ZeroInt()
+		if len(vestingCoins) > 0 {
+			vestingAmt = vestingCoins[j].Amount
+		}
+
+		delVestingAmt := sdk.ZeroInt()
+		if len(bva.DelegatedVesting) > 0 {
+			delVestingAmt = bva.DelegatedVesting[k].Amount
+		}
+
+		// Panic if the delegation amount is zero or if the base coins does not
+		// exceed the desired delegation amount.
+		if coin.Amount.IsZero() || baseAmt.LT(coin.Amount) {
+			panic("delegation attempt with zero coins or insufficient funds")
+		}
+
+		// compute x and y per the specification, where:
+		// X := min(max(V - DV, 0), D)
+		// Y := D - X
+		x := sdk.MinInt(sdk.MaxInt(vestingAmt.Sub(delVestingAmt), sdk.ZeroInt()), coin.Amount)
+		y := coin.Amount.Sub(x)
+
+		if !x.IsZero() {
+			xCoin := sdk.NewCoin(coin.Denom, x)
+			bva.DelegatedVesting = bva.DelegatedVesting.AddCoinByDenom(xCoin)
+		}
+
+		if !y.IsZero() {
+			yCoin := sdk.NewCoin(coin.Denom, y)
+			bva.DelegatedFree = bva.DelegatedFree.AddCoinByDenom(yCoin)
+		}
+
+		bva.Coins = bva.Coins.SubCoinByDenom(coin)
```

Comment:
This is weird, we're modifying the account itself (passed by reference). Should the caller save it afterwards? We don't actually save it in the bank implementation, insofar as I can see...

Reply:
I guess we do set the returned coins, so it works. I still think this is confusing - either the function should modify the account and require that the account be written, or not modify the account and require an explicit `setCoins`, not do both.